### PR TITLE
Fixed the GELU gradient problem mentioned in #107876.

### LINKED ISSUE #107876: Gradients of GELU inconsistent on CPU, work on GPU
### 🐛 Describe the bug

Gradients of layers before a GELU followed by two transposes are inconsistent when computed on CPU, but work with CUDA. This issue does not occur when using ReLU.

To reproduce, use the code snippet below.

To observe the inconsistency with the CPU, run this script several times. The gradients are not fully random. Some values are reoccuring. Commenting in the `.cuda()` calls should produce reproducible results.

```python
import os
import random

import numpy as np
import torch
import torch.nn as nn

def set_seed(seed: int):
    np.random.seed(seed)
    random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False
    os.environ["PYTHONHASHSEED"] = str(seed)


set_seed(42)

test_in = torch.ones(32, 100, 1000)
test_target = torch.ones(32, 100, 1000)

first_layer = nn.Linear(1000, 1000)
second_layer = nn.Linear(100, 100)

with torch.no_grad():
    first_layer.weight = nn.Parameter(torch.ones_like(first_layer.weight))
    second_layer.weight = nn.Parameter(torch.ones_like(second_layer.weight))

# COMMENT IN FOR REPRODUCIBLE GRADIENTS
# test_in = test_in.cuda()
# test_target = test_target.cuda()
# first_layer = first_layer.cuda()
# second_layer = second_layer.cuda()

loss = torch.mean(
    (
        second_layer(
            nn.functional.gelu(first_layer(test_in)).transpose(-1, -2)
        ).transpose(-1, -2)
        - test_target
    )
    ** 2.0
)
loss.backward()

print(first_layer.weight.grad)
```

Example output when run on CPU:
```python
# First run
tensor([[2.1599e+33, 2.1599e+33, 2.1599e+33,  ..., 2.1599e+33, 2.1599e+33,
         2.1599e+33],
        [3.2001e+06, 3.2001e+06, 3.2001e+06,  ..., 3.2001e+06, 3.2001e+06,
         3.2001e+06],
        [3.2000e+06, 3.2000e+06, 3.2000e+06,  ..., 3.2000e+06, 3.2000e+06,
         3.2000e+06],
        ...,
        [3.2000e+06, 3.2000e+06, 3.2000e+06,  ..., 3.2000e+06, 3.2000e+06,
         3.2000e+06],
        [3.2000e+06, 3.2000e+06, 3.2000e+06,  ..., 3.2000e+06, 3.2000e+06,
         3.2000e+06],
        [3.2000e+06, 3.2000e+06, 3.2000e+06,  ..., 3.2000e+06, 3.2000e+06,
         3.2000e+06]])

# Second run
tensor([[2.0544e+37, 2.0544e+37, 2.0544e+37,  ..., 2.0544e+37, 2.0544e+37,
         2.0544e+37],
        [2.0000e+02, 2.0000e+02, 2.0000e+02,  ..., 2.0000e+02, 2.0000e+02,
         2.0000e+02],
        [2.0000e+02, 2.0000e+02, 2.0000e+02,  ..., 2.0000e+02, 2.0000e+02,
         2.0000e+02],
        ...,
        [2.0000e+02, 2.0000e+02, 2.0000e+02,  ..., 2.0000e+02, 2.0000e+02,
         2.0000e+02],
        [2.0000e+02, 2.0000e+02, 2.0000e+02,  ..., 2.0000e+02, 2.0000e+02,
         2.0000e+02],
        [2.0000e+02, 2.0000e+02, 2.0000e+02,  ..., 2.0000e+02, 2.0000e+02,
         2.0000e+02]])

# Third run
tensor([[199.4996, 199.4996, 199.4996,  ..., 199.4996, 199.4996, 199.4996],
        [200.0014, 200.0014, 200.0014,  ..., 200.0014, 200.0014, 200.0014],
        [199.9969, 199.9969, 199.9969,  ..., 199.9969, 199.9969, 199.9969],
        ...,
        [200.0000, 200.0000, 200.0000,  ..., 200.0000, 200.0000, 200.0000],
        [199.9984, 199.9984, 199.9984,  ..., 199.9984, 199.9984, 199.9984],
        [199.9980, 199.9980, 199.9980,  ..., 199.9980, 199.9980, 199.9980]])

# Fourth run
tensor([[2.0544e+37, 2.0544e+37, 2.0544e+37,  ..., 2.0544e+37, 2.0544e+37,
         2.0544e+37],
        [3.2001e+06, 3.2001e+06, 3.2001e+06,  ..., 3.2001e+06, 3.2001e+06,
         3.2001e+06],
        [3.2000e+06, 3.2000e+06, 3.2000e+06,  ..., 3.2000e+06, 3.2000e+06,
         3.2000e+06],
        ...,
        [3.2000e+06, 3.2000e+06, 3.2000e+06,  ..., 3.2000e+06, 3.2000e+06,
         3.2000e+06],
        [3.2000e+06, 3.2000e+06, 3.2000e+06,  ..., 3.2000e+06, 3.2000e+06,
         3.2000e+06],
        [3.2000e+06, 3.2000e+06, 3.2000e+06,  ..., 3.2000e+06, 3.2000e+06,
         3.2000e+06]])
```

Example output when run with CUDA (always the same):
```python
tensor([[19999.9629, 19999.9629, 19999.9629,  ..., 19999.9629, 19999.9629,
         19999.9629],
        [20000.0469, 20000.0469, 20000.0469,  ..., 20000.0469, 20000.0469,
         20000.0469],
        [19999.8125, 19999.8125, 19999.8125,  ..., 19999.8125, 19999.8125,
         19999.8125],
        ...,
        [20000.0020, 20000.0020, 20000.0020,  ..., 20000.0020, 20000.0020,
         20000.0020],
        [19999.9512, 19999.9512, 19999.9512,  ..., 19999.9512, 19999.9512,
         19999.9512],
        [19999.8398, 19999.8398, 19999.8398,  ..., 19999.8398, 19999.8398,
         19999.8398]], device='cuda:0')
```

### Versions

```Collecting environment information...
PyTorch version: 1.11.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.8.10 (default, May 26 2023, 14:05:08)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-42-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          8
On-line CPU(s) list:             0-7
Thread(s) per core:              1
Core(s) per socket:              1
Socket(s):                       8
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) Gold 5218 CPU @ 2.30GHz
Stepping:                        7
CPU MHz:                         2299.998
BogoMIPS:                        4599.99
Virtualization:                  VT-x
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       256 KiB
L1i cache:                       256 KiB
L2 cache:                        32 MiB
L3 cache:                        128 MiB
NUMA node0 CPU(s):               0-7
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Mitigation; TSX disabled
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat umip pku ospke avx512_vnni md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] torch==1.11.0
[conda] Could not collect
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### PR DESCRIPTION
**Pull request by tringwald**
Fixes #107876. Apparently the `ideep::tensor` had some trouble with the transposed input tensor. With these changes, the results on CPU and CUDA are now equal (besides numerical issues).

cc @yanbing-j

### CODE REVIEW

**Comment by yanbing-j** [comment]
Hi @tringwald , the root cause of GELU gradient issue is the share buffer issue in ideep. Please refer to https://github.com/pytorch/pytorch/pull/108038. Thanks!

**Comment by tringwald** [comment]
Great find! I guess this supersedes this PR then.
